File: internal/api/brand_theme_test.go
package api

import "testing"

func TestBrandColorPresetOptionsMatchSettingsCatalog(t *testing.T) {
	want := []string{
		"aurora", "harbor", "sunset", "botanical", "meadow", "citrus",
		"gold", "orange", "red", "pink", "violet", "blue", "cyan", "teal", "green", "lime",
	}
	if len(brandColorPresetOptions) != len(want) {
		t.Fatalf("brand preset count = %d, want %d", len(brandColorPresetOptions), len(want))
	}
	for index := range want {
		if brandColorPresetOptions[index] != want[index] {
			t.Fatalf("brand preset %d = %q, want %q", index, brandColorPresetOptions[index], want[index])
		}
	}
}

func TestNormalizeBrandColorPresetFallsBackToBlue(t *testing.T) {
	if got := normalizeBrandColorPreset(" Teal "); got != "teal" {
		t.Fatalf("normalized preset = %q, want teal", got)
	}
	if got := normalizeBrandColorPreset("custom-css"); got != defaultBrandColorPreset {
		t.Fatalf("invalid preset = %q, want %q", got, defaultBrandColorPreset)
	}
}

func TestNormalizeUserBrandColorPresetAllowsFollowingPlatform(t *testing.T) {
	if got, valid := normalizeUserBrandColorPreset("  "); !valid || got != "" {
		t.Fatalf("empty user preset = %q, valid=%v; want empty and valid", got, valid)
	}
	if got, valid := normalizeUserBrandColorPreset(" Teal "); !valid || got != "teal" {
		t.Fatalf("official user preset = %q, valid=%v; want teal and valid", got, valid)
	}
	if got, valid := normalizeUserBrandColorPreset(" Ruby "); valid || got != "ruby" {
		t.Fatalf("hidden user preset = %q, valid=%v; want ruby and invalid", got, valid)
	}
	if got, valid := normalizeUserBrandColorPreset("custom-css"); valid || got != "custom-css" {
		t.Fatalf("custom user preset = %q, valid=%v; want invalid", got, valid)
	}
}

func TestValidateConfigValuesRejectsUnknownBrandColorPreset(t *testing.T) {
	for _, invalid := range []string{"custom-css", "ruby"} {
		if _, err := validateConfigValues(map[string]any{siteBrandColorPresetKey: invalid}); err == nil {
			t.Fatalf("expected brand color preset %q to be rejected", invalid)
		}
	}
	values, err := validateConfigValues(map[string]any{siteBrandColorPresetKey: "red"})
	if err != nil {
		t.Fatalf("validate official brand color preset: %v", err)
	}
	if values[siteBrandColorPresetKey] != "red" {
		t.Fatalf("validated preset = %q, want red", values[siteBrandColorPresetKey])
	}
}

File: web/src/app/brand-theme.test.ts
import { beforeEach, describe, expect, it } from 'vitest'
import { applySiteBrandColorPreset, applyUserBrandColorPreference, brandColorPresets, brandColorUsesDarkForeground, brandThemeSwatchBackground, brandThemeSwatchColors, clearActiveUserBrandColorPreference, defaultBrandColorPreset, normalizeBrandColorPreset, normalizeUserBrandColorPreference } from './brand-theme'

describe('brand theme presets', () => {
  beforeEach(() => {
    localStorage.clear()
    delete document.documentElement.dataset.brandTheme
  })

  it('uses one accepted catalog for the curated picker', () => {
    expect(brandColorPresets).toEqual([
      'aurora',
      'harbor',
      'sunset',
      'botanical',
      'meadow',
      'citrus',
      'gold',
      'orange',
      'red',
      'pink',
      'violet',
      'blue',
      'cyan',
      'teal',
      'green',
      'lime',
    ])
  })

  it('falls back to the default for unknown values', () => {
    expect(normalizeBrandColorPreset(' Teal ')).toBe('teal')
    expect(normalizeBrandColorPreset('custom-css')).toBe(defaultBrandColorPreset)
  })

  it('uses a dark foreground for the retained bright solid scale', () => {
    expect(brandColorPresets.filter(brandColorUsesDarkForeground)).toEqual(['lime'])
  })

  it('describes composite and single-color swatches consistently', () => {
    expect(brandThemeSwatchColors('aurora')).toEqual(['#3b6fe8', '#7867d9', '#2e9eaa', '#e7a63a'])
    expect(brandThemeSwatchColors('botanical')?.[0]).toBe('#0d4336')
    expect(brandThemeSwatchColors('blue')).toBeNull()
    expect(brandThemeSwatchBackground('aurora')).toBe('#3b6fe8')
    expect(brandThemeSwatchBackground('blue')).toBe('var(--blue-9)')
  })

  it('keeps an empty user preference as platform inheritance', () => {
    expect(normalizeUserBrandColorPreference('')).toBe('')
    expect(normalizeUserBrandColorPreference(' Teal ')).toBe('teal')
    expect(normalizeUserBrandColorPreference(' Ruby ')).toBe('')
    expect(normalizeUserBrandColorPreference('custom-css')).toBe('')
  })

  it('applies user preference before site preference and restores the site preference', () => {
    applySiteBrandColorPreset('blue')
    applyUserBrandColorPreference('usr_theme', 'teal')
    expect(document.documentElement.dataset.brandTheme).toBe('teal')

    applySiteBrandColorPreset('red')
    expect(document.documentElement.dataset.brandTheme).toBe('teal')

    applyUserBrandColorPreference('usr_theme', '')
    expect(document.documentElement.dataset.brandTheme).toBe('red')

    clearActiveUserBrandColorPreference()
    expect(document.documentElement.dataset.brandTheme).toBe('red')
  })
})
